### ISSUE TITLE
Screen space reflection with roughness does not blur across object edges

### ISSUE THREAD

**Issue by cjxgm** [godotengine/godot#25112]
**Godot version:** 3.1 beta1

**OS/device including version:** Windows, Linux

**Issue description:**

![image](https://user-images.githubusercontent.com/490327/51411424-b7929680-1ba2-11e9-9d00-6ea5f707f272.png)
As shown in the image, the ground has a roughness of 0 (left) and 0.3 (right) with SSR enabled.
Details *inside* objects get blurred when roughness is large (as shown in green circles) as expected, but object edges does not get blurred (as shown in red circles).

Blurred object edges are expected.

**Steps to reproduce:**
Enable SSR in default_env, add a ground mesh, add a spatial material with metalness set to 1 and roughness set to non-zero. Add any mesh to be reflected.

**Comment by cjxgm** [comment]
This is definitely not the limitations of screen space reflections, but the limitations of the particular implementation (conetracing?) of godot.

For example, in Blender Eevee, SSR is correct:
![image](https://user-images.githubusercontent.com/490327/51422506-8ccc3080-1bea-11e9-93a8-ff70171b21b9.png)

**Comment by puchik** [comment]
I've done some fiddling and hopefully this can be of some use:

When you progressively increase roughness, notice how a faint blur starts to appear around the objects (around 4 or 5 it becomes noticeable). It's almost as if roughness works BUT the non-rough version is being drawn below the rough one.

![compare](https://user-images.githubusercontent.com/48544263/55666118-c29eaf80-57fe-11e9-8f87-8eac272ec37c.jpg)

Even to achieve this effect though, you need to have another object (or have "max steps" high enough so the sky gets reflected) behind it to work. Otherwise, it is completely sharp on the edges.

![compare2](https://user-images.githubusercontent.com/48544263/55666155-3771e980-57ff-11e9-9d2f-b14a90de228d.jpg)

You can also pitch the camera down, and once you get to just barely on top of reflective surface...:
1. The roughness seems to **actually** work properly.
2. Once you remove the object behind it, you get artifacts (if max steps is high) OR completely sharp reflection (if max steps is not high enough). The lower you go, the worse this gets.
   - Notice how in the artifacts, it seems to alternate between sharp and blurry.


![compare3](https://user-images.githubusercontent.com/48544263/55666176-861f8380-57ff-11e9-8cb1-1ddf842c5cee.jpg)

Here as well, in alternating artifact lines you can see how the red cube's reflection is being blurred out 

Basically, if you have an object behind it and you're at a low angle, it works (except for the very faint outline you can still see in the rightmost picture here).

#18395 mentions artifacts that sound similar but the video is gone so I'm not sure if it's the same. In this case disabling roughness doesn't help, but may cause the lines to be less visible sometimes.

Just some things I observed and am hoping can be useful!

.zip of all the pics
[ssr_roughness.zip](https://github.com/godotengine/godot/files/3050235/ssr_roughness.zip)

**Comment by Zireael07** [comment]
It's the same artifacts as in my issue, and since you said the video is gone, I will close it in favor of this one!

**Comment by adriansnetlis** [comment]
![ssr-glossy-color](https://user-images.githubusercontent.com/6423934/62118207-cc1bbe00-b2ac-11e9-8b2e-8da7c4eaa3e0.png)
![ssr-rough-nocolor](https://user-images.githubusercontent.com/6423934/62118209-cd4ceb00-b2ac-11e9-999f-a68992a4b011.png)
Another thing I wanted to note is that, as visible in the images, with increased roughness the reflections loose color. At higher roughness values the reflections can become totally unusable.

**Comment by Calinou** [comment]
@adriansnetlis Please open a separate issue for this, as this appears to be unrelated to the issue reported in the first post.

**Comment by adriansnetlis** [comment]
#30989

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by puchik** [comment]
![image](https://user-images.githubusercontent.com/48544263/94370687-3ba25e80-00a6-11eb-8f31-16fdc3a0ef5b.png)

Still reproducible in 3.2.3

[ssr-edge-fade-broken.zip](https://github.com/godotengine/godot/files/5288679/ssr-edge-fade-broken.zip)

**Comment by Calinou** [comment]
As per https://github.com/godotengine/godot/issues/49402#issuecomment-856240838, this is fixed in the `master` branch but I'm not sure if this can be fixed in the `3.x` branch without rewriting SSR (which would take too much time).

**Comment by puchik** [comment]
You could apply a post process blur to the screen space reflection before applying it. I've used a change like that in a project and some custom builds, but it's kind of hacky and not really physically accurate (but it was good enough for me). I'll probably put it up as a patch or something but I don't know if that's a good solution for the actual engine source code. 

Maybe someone else has some fancy (not quick'n'dirty'hacky) post process blur ideas?
